src/types.md

```diff
@@ -0,0 +1,183 @@
+[the answer about the observer pattern](./codebase.md#the-c-observer-pattern-is-hard-in-rust-what-to-do) and especially
+[the second option described there](./codebase.md#option-2-drive-the-objects-from-the-code-not-the-other-way-round).)
+
+Even in Rust, though, it's still often the best choice to have complex
+behavior as part of types within `impl` blocks. You can still do that -
+but don't _store_ references. Instead, pass them into each function call.
+
+```mermaid
+flowchart LR
+    subgraph Shared functionality
+    important[Important Shared Object]
+    end
+    subgraph ObjectA
+    methodA[Method]
+    methodA-. Acts on shared object.->important
+    end
+    subgraph ObjectB
+    methodB[Method]
+    methodB-. Acts on shared object.->important
+    end
+    main --> ObjectA
+    main --> ObjectB
+    main --> important
+    main-. Passes reference to shared object.-> methodA
+    main-. Passes reference to shared object.-> methodB
+```
+
+Instead of:
+
+```rust
+struct ImportantSharedObject;
+
+struct ObjectA<'a> {
+    important_shared_object: &'a mut ImportantSharedObject,
+}
+
+impl<'a> ObjectA<'a> {
+    fn new(important_shared_object: &'a mut ImportantSharedObject) -> Self {
+        Self {
+            important_shared_object
+        }
+    }
+    fn do_something(&mut self) {
+        // act on self.important_shared_object
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut a = ObjectA::new(&mut shared_thingy);
+    a.do_something();
+}
+```
+
+Do this:
+
+```rust
+struct ImportantSharedObject;
+
+struct ObjectA;
+
+impl ObjectA {
+    fn new() -> Self {
+        Self
+    }
+    fn do_something(&mut self, important_shared_object: &mut ImportantSharedObject) {
+        // act on important_shared_object
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut a = ObjectA::new();
+    a.do_something(&mut shared_thingy);
+}
+```
+
+(happily this also gets rid of named lifetime parameters.)
+
+If you have a hundred such shared objects, you probably don't want a
+hundred function parameters. So it's usual to bundle them up into
+a context structure which can be passed into each function call:
+
+```rust
+struct ImportantSharedObject;
+struct AnotherImportantObject;
+
+struct Ctx<'a> {
+    important_shared_object: &'a mut ImportantSharedObject,
+    another_important_object: &'a mut AnotherImportantObject,
+}
+
+struct ObjectA;
+
+impl ObjectA {
+    fn new() -> Self {
+        Self
+    }
+    fn do_something(&mut self, ctx: &mut Ctx) {
+        // act on ctx.important_shared_object and ctx.another_important_thing
+    }
+}
+
+fn main() {
+    let mut shared_thingy = ImportantSharedObject;
+    let mut another_thingy = AnotherImportantObject;
+    let mut ctx = Ctx {
+        important_shared_object: &mut shared_thingy,
+        another_important_object: &mut another_thingy,
+    };
+    let mut a = ObjectA::new();
+    a.do_something(&mut ctx);
+}
+```
+
```

Comment:
Here, I think it's important to show the definition of `Ctx`, but you may again be able to leave out the definition of `ObjectA`.